plan/util.go

```diff
@@ -45,3 +46,60 @@ func (a *AggregateFuncExtractor) Leave(n ast.Node) (ast.Node, bool) {
 	}
 	return n, true
 }
+
+// logicalSchemaProducer stores the schema for the logical plans who can produce schema directly.
+type logicalSchemaProducer struct {
+	schema *expression.Schema
+	baseLogicalPlan
+}
+
+// Schema implements the Plan.Schema interface.
+func (s *logicalSchemaProducer) Schema() *expression.Schema {
+	if s.schema == nil {
+		s.schema = expression.NewSchema()
+	}
+	return s.schema
+}
+
+// SetSchema implements the Plan.SetSchema interface.
+func (s *logicalSchemaProducer) SetSchema(schema *expression.Schema) {
+	s.schema = schema
+}
+
+// physicalSchemaProducer stores the schema for the physical plans who can produce schema directly.
+type physicalSchemaProducer struct {
+	schema *expression.Schema
+	basePhysicalPlan
+}
+
+// Schema implements the Plan.Schema interface.
+func (s *physicalSchemaProducer) Schema() *expression.Schema {
+	if s.schema == nil {
+		s.schema = expression.NewSchema()
+	}
+	return s.schema
+}
+
+// SetSchema implements the Plan.SetSchema interface.
+func (s *physicalSchemaProducer) SetSchema(schema *expression.Schema) {
+	s.schema = schema
+}
+
+// baseSchemaProducer stores the schema for the base plans who can produce schema directly.
+type baseSchemaProducer struct {
+	schema *expression.Schema
```

Comment:
put `schema` into basePlan is enough?

Reply:
If put schema into basePlan, we can set schema for every plan.